Automattic/woocommerce-payments
# Fix bug with adding order metadata fails and refund is not possible

Fixes #2656 

**Bug background and investigation**

It's not possible to click the _Refund with WooCommerce payments_ button for some orders.

After an initial investigation, it seemed to be due to missing `_intent_id` from the `wp_postmeta` table. It ends up like this one for different reasons, all of those are related to errors in the method below when different plugins are connected:
https://github.com/Automattic/woocommerce-payments/blob/1c1df7bb704b18758312ba57f100366af9564e11/includes/class-wc-payment-gateway-wcpay.php#L1151

I've made an additional investigation and discovered next:
- The code that is responsible for showing the Refund checks if it can refund the order via the gateway.
- In case of WooCommerce Payments the check is implemented here
https://github.com/Automattic/woocommerce-payments/blob/1c1df7bb704b18758312ba57f100366af9564e11/includes/class-wc-payment-gateway-wcpay.php#L1244

Eventually, it means that the not only `_intent_id` meta item missing is the reason for the bug, but also `_charge_id`. 

**How to reproduce the bug**

I was able to reproduce the missing button on my local after deleting the meta value by key `_charge_id` in the `wp_postmeta` table.

#### Changes proposed in this Pull Request

One of the solutions that were discussed in the slack channel was to change the order of how the code blocks are executed in `WC_Payment_Gateway_WCPay::attach_intent_info_to_order()`.
It might not provide 100% coverage for this bug, but given that covering 100% of the cases here might be virtually impossible it might be good enough. 
Another one was to `try/catch` the errors from the `payment_complete` method. If some 3rd party plugin will throw an error, the code will be executed despite the failure.
I believe that a combination of those two changes will solve the issue for new orders and meta items will be stored properly and then the button would be available.

**Important**: old orders won't be fixes as they are already missing the meta value by key `_charge_id` in the `wp_postmeta` table.

#### Testing instructions

1. Go to the shop as a shopper, add something to the cart, do a checkout, and complete the order.
2. Check that both `_intent_id` and `_charge_id` meta values are present in `wp_postmeta` table.
3. Go to admin as a merchant, find the order and try to refund it. The button should appear. 
4. Do the refund, check that everything works properly.
5. Enable manual charge capture in the WCPay settings. Repeat steps 1-4.
* Unit tests should pass.

-------------------


**Post merge**
